File: src/LeetCode/medium/74-Search2DMatrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */

// ** Diagonal search from top right - O(m + n)**
var searchMatrix = function (matrix, target) {
    // start from top right corner - if smaller move left - if larger move down
    let i = 0,
        j = matrix[0].length - 1;

    while (i < matrix.length && j >= 0) {
        if (matrix[i][j] == target) return true;
        if (matrix[i][j] < target) {
            i++;
        } else {
            j--;
        }
    }

    return false;
};

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
// O(n) = log(m*n)
var searchMatrix = function (matrix, target) {

    // binary search and treat the 2d matrix as an ascending sorted array
    let start = 0,
        end = (matrix.length * matrix[0].length) - 1;

    while (start <= end) {
        let mid = Math.floor((start + end) / 2);
        let midNumber = matrix[Math.floor(mid / matrix[0].length)][mid % matrix[0].length];

        if (midNumber == target) return true;
        else if (midNumber > target) {
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    }

    return false;
};

File: LeetCode/top150/74-search-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        # matrix_to_list = []

        # for row in matrix:
        #     for column in row:
        #         matrix_to_list.append(column)
        
        # i, j = 0, len(matrix_to_list)-1

        # while i <= j:
        #     mid = (i + j) // 2
        #     if matrix_to_list[mid] == target:
        #         return True
        #     elif matrix_to_list[mid] < target:
        #         i = mid + 1
        #     else:
        #         j = mid - 1
        
        # return False
        # TC: O(log(m*n)) SC: O(m*n)

        top = 0
        bot = len(matrix) - 1

        while top <= bot:
            mid = (top + bot) // 2
            if matrix[mid][0] > target:
                bot = mid - 1
            elif matrix[mid][len(matrix[0])-1] < target:
                top = mid + 1
            else:
                break
        
        row = (top + bot) // 2 # Found row

        l = 0
        r = len(matrix[0]) - 1

        while l <= r:
            mid = (l + r) // 2
            if matrix[row][mid] == target:
                return True
            elif matrix[row][mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        
        return False

    